### PR TITLE
Revise "rnn-scratch_vn" - Phần 2

### LINKED ISSUE #1702: Revise "rnn-scratch_vn" - Phần 2
Phần này được dịch bởi: **tranyenthy**
Nếu bạn đã dịch phần này, vui lòng bỏ qua việc revise.

### PR DESCRIPTION
**Pull request by cuongvng**
Close #1702

### CODE REVIEW

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 1]
```diff
@@ -234,10 +234,10 @@ For the beginning of the sequence, we only update the hidden state.
...
 -->
 
-Trước tiên cần giải thích về hàm dự đoán để chúng ta có thể thường xuyên kiểm tra trong quá trình huấn luyện.
-Hàm này dự đoán các ký tự `num_predicts` tiếp theo dựa trên `prefix` (một chuỗi chứa một vài ký tự).
-Để bắt đầu chuỗi, ta cập nhật chỉ trạng thái ẩn.
-Sau đó, ta bắt đầu tạo ra các ký tự mới và ban hành chúng.
+Trước tiên chúng ta giải thích hàm dự đoán để có thể kiểm tra thường xuyên trong quá trình huấn luyện.
+Hàm này dự đoán `num_predicts` ký tự tiếp theo dựa trên `prefix` (một chuỗi chứa một vài ký tự).
+Khi bắt đầu chuỗi, ta chỉ cập nhật trạng thái ẩn.
+Sau đó bắt đầu tạo ra các ký tự mới.
```
Comment (lkhphuc) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:240]:
```suggestion
Ở các ký tự đầu tiên trong chuỗi, ta chỉ cập nhật trạng thái ẩn rồi sau đó mới bắt đầu tạo ra các ký tự mới.
```

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 2]
```diff
@@ -276,26 +276,25 @@ predict_ch8('time traveller ', 10, model, vocab, ctx)
...
 As mentioned in :numref:`sec_numerical_stability`, it might result in numerical instability, e.g., the gradients may either explode or vanish, when $T$ is large.
 Therefore, RNN models often need extra help to stabilize the training.
 -->
 
-Đối với chuỗi có độ dài $T$, ta tính toán gradient theo các bước thời gian $T$ này trong một vòng lặp, dẫn đến một dây chuyền các tích ma trận có độ dài $\mathcal{O}(T)$ trong quá trình lan truyền ngược.
-Như đã đề cập trong :numref:`sec_numerical_stability`, việc này có thể dẫn đến mất ổn định số, ví dụ: các gradient có thể phát nổ hoặc biến mất, khi $T$ lớn.
-Do đó, các mô hình RNN thường cần thêm trợ giúp để ổn định việc huấn luyện.
+Với chuỗi độ dài $T$, trong một vòng lặp lan truyền ngược ta tính toán gradient qua $T$ bước thời gian, dẫn đến một chuỗi các tích ma trận có độ phức tạp $\mathcal{O}(T)$.
+Như đã đề cập trong :numref:`sec_numerical_stability`, việc này có thể dẫn đến mất ổn định số học, qua hiện tượng bùng nổ hoặc tiêu biến gradient khi $T$ lớn.
```
Comment (lkhphuc) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:288]:
```suggestion
Như đã đề cập trong :numref:`sec_numerical_stability`,  khi $T$ lớn việc này có thể dẫn đến mất ổn định số học, qua hiện tượng bùng nổ hoặc tiêu biến gradient.
```

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 3]
```diff
@@ -309,9 +308,9 @@ This is both a curse and a blessing.
 A curse since it limits the speed of making progress, whereas a blessing since it limits the extent to which things can go wrong if we move in the wrong direction.
 -->
 
-Trong trường hợp này, ta có thể giả định một cách an toàn rằng bằng cách cập nhật vector trọng số theo $\eta \cdot \mathbf{g}_t$, sự thay đổi sẽ không lớn hơn $L \eta \|\mathbf{g}_t\|$.
-Đây là cả một lời nguyền và một phước lành.
-Một lời nguyền là vì nó giới hạn tốc độ tiến bộ, trong khi việc nó là một phước lành là bởi nó hạn chế mức độ sai lệnh trong trường hợp chúng ta đi sai hướng.
+Trong trường hợp này, có thể nói khi cập nhật vector trọng số theo $\eta \cdot \mathbf{g}_t$, sự thay đổi sẽ không lớn hơn $L \eta \|\mathbf{g}_t\|$.
+Điều này vừa có lợi vừa có hại.
+Có hại ở chỗ tốc độ tối ưu bị giới hạn, có lợi ở chỗ mức độ sai lệnh khi tối ưu sai hướng cũng bị hạn chế.
```
Comment (lkhphuc) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:313]:
```suggestion
Có hại ở chỗ tốc độ tối ưu bị giới hạn, có lợi ở chỗ mức độ sai lệch khi tối ưu sai hướng cũng bị hạn chế.
```

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 4]
```diff
@@ -321,11 +320,10 @@ In this case such an approach may appear entirely unwarranted.
 One alternative is to clip the gradients by projecting them back to a ball of a given radius, say $\theta$ via
 -->
 
-Đôi khi gradient có thể khá lớn và thuật toán tối ưu có thể không hội tụ.
-Ta có thể giải quyết vấn đề này bằng cách giảm tốc độ học $\eta$ hoặc bằng một số thủ thuật bậc cao khác.
-Nhưng điều gì sẽ xảy ra nếu chúng ta hiếm khi nhận được gradient lớn?
-Trong trường hợp này, cách tiếp cận như vậy không được bảo đảm hoàn toàn.
-Một cách khác là cắt bớt các gradient bằng cách chiếu chúng trở lại một quả cầu với bán kính $\theta$ thông qua:
+Đôi khi gradient có thể khá lớn và do đó thuật toán tối ưu không hội tụ.
+Có thể giải quyết vấn đề này bằng cách giảm tốc độ học $\eta$ hoặc một số thủ thuật cao cấp khác.
```
Comment (lkhphuc) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:324]:
```suggestion
Vấn đề này có thể được giả quyết bằng cách giảm tốc độ học $\eta$ hoặc sử dụng một số thủ thuật liên quan tới đạo hàm bậc cao hơn.
```
high-order chỉ second/third order derivative ấy.

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 5]
```diff
@@ -337,19 +335,17 @@ Gradient clipping provides a quick fix to the gradient exploding.
 While it does not entirely solve the problem, it is one of the many techniques to alleviate it.
 -->
 
-Bằng cách làm như vậy, ta biết rằng chuẩn độ dốc không bao giờ vượt quá $\theta$ và độ dốc được cập nhật hoàn toàn phù hợp với hướng ban đầu $\mathbf{g}$.
-Nó cũng có tác dụng phụ tích cực là hạn chế ảnh hưởng của bất kỳ minibatch nào (và bên trong nó là bất kỳ mẫu nào) có thể tác động lên các vector trọng số.
-Điều này mang lại một độ mạnh mẽ nhất định cho mô hình.
-Cắt bớt gradient là một phương án sửa chữa nhanh chóng cho vấn đề phát nổ gradient.
-Mặc dù nó không hoàn toàn giải quyết vấn đề, nhưng là một trong nhiều kỹ thuật để giảm bớt vấn đề đó.
+Như vậy chuẩn của gradient sẽ không vượt quá $\theta$ và gradient sau khi gọt sẽ cùng hướng gradient $\mathbf{g}$ ban đầu.
+Gọt gradient có tác dụng phụ tích cực là hạn chế ảnh hưởng quá lớn của bất kỳ minibatch nào (hoặc bất kỳ mẫu nào) lên các trọng số, làm cho mô hình mạnh mẽ hơn.
```
Comment (lkhphuc) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:339]:
```suggestion
Gọt gradient có tác dụng phụ tích cực là hạn chế ảnh hưởng quá lớn của bất kỳ minibatch nào (hoặc bất kỳ mẫu nào) lên các trọng số, làm cho mô hình ổn định hơn.
```

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 6]
```diff
@@ -337,19 +335,17 @@ Gradient clipping provides a quick fix to the gradient exploding.
...
+Đây là một trong các kỹ thuật có thể nhanh chóng giảm thiểu bùng nổ gradient, dù không hoàn toàn giải quyết được.
```
Comment (lkhphuc) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:340]:
```suggestion
Dù không giải quyết được hoàn toàn vấn đề, đây là một kỹ thuật đơn giản để làm giảm nhẹ vấn đề bùng nổ gradient.
```

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md (hunk 2 above)
Comment (cuongvng) [reply chapter_recurrent-neural-networks/rnn-scratch_vn.md:287]:
```suggestion
Như đã đề cập trong :numref:`sec_numerical_stability`, khi $T$ lớn việc này có thể dẫn đến mất ổn định số học, qua hiện tượng bùng nổ hoặc tiêu biến gradient.
```
em sửa 1 spacing

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 7]
```diff
@@ -234,10 +234,9 @@ For the beginning of the sequence, we only update the hidden state.
 After that we begin generating new characters and emitting them.
 -->
 
-Trước tiên cần giải thích về hàm dự đoán để chúng ta có thể thường xuyên kiểm tra trong quá trình huấn luyện.
-Hàm này dự đoán các ký tự `num_predicts` tiếp theo dựa trên `prefix` (một chuỗi chứa một vài ký tự).
-Để bắt đầu chuỗi, ta cập nhật chỉ trạng thái ẩn.
-Sau đó, ta bắt đầu tạo ra các ký tự mới và ban hành chúng.
+Trước tiên chúng ta giải thích hàm dự đoán mà thường xuyên được dùng để kiểm tra trong quá trình huấn luyện.
```
Comment (rootonchair) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:237]:
```suggestion
Trước tiên chúng ta giải thích hàm dự đoán thường xuyên được dùng để kiểm tra trong quá trình huấn luyện.
```

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 8]
```diff
@@ -263,9 +262,9 @@ Given that we did not train the network, it will generate nonsensical prediction
 We initialize it with the sequence `traveller ` and have it generate 10 additional characters.
 -->
 
-Ta kiểm tra hàm `predict_rnn` trước tiên.
-Vì mạng không được huấn luyện, nó sẽ tạo ra các dự đoán vô nghĩa.
-Ta khởi tạo mạng với chuỗi `traveller` và để nó tạo thêm 10 ký tự.
+Ta kiểm tra hàm `predict_ch8` trước.
```
Comment (rootonchair) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:265]:
```suggestion
Ta chạy thử hàm `predict_ch8` trước.
```

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 9]
```diff
@@ -263,9 +262,9 @@ Given that we did not train the network, it will generate nonsensical prediction
...
+Lúc này đầu ra sẽ là các dự đoán vô nghĩa do mạng chưa được huấn luyện.
+Ta khởi tạo mạng với chuỗi `traveller` và tạo ra thêm 10 ký tự.
```
Comment (rootonchair) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:267]:
```suggestion
Ta khởi tạo mạng với chuỗi `traveller` và cho nó tạo ra thêm 10 ký tự.
```

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 10]
```diff
@@ -276,26 +275,25 @@ predict_ch8('time traveller ', 10, model, vocab, ctx)
...
 <!--
 For a sequence of length $T$, we compute the gradients over these $T$ timesteps in an iteration, which results in a chain of matrix-products with length $\mathcal{O}(T)$ during backpropagating.
 As mentioned in :numref:`sec_numerical_stability`, it might result in numerical instability, e.g., the gradients may either explode or vanish, when $T$ is large.
 Therefore, RNN models often need extra help to stabilize the training.
 -->
 
-Đối với chuỗi có độ dài $T$, ta tính toán gradient theo các bước thời gian $T$ này trong một vòng lặp, dẫn đến một dây chuyền các tích ma trận có độ dài $\mathcal{O}(T)$ trong quá trình lan truyền ngược.
-Như đã đề cập trong :numref:`sec_numerical_stability`, việc này có thể dẫn đến mất ổn định số, ví dụ: các gradient có thể phát nổ hoặc biến mất, khi $T$ lớn.
-Do đó, các mô hình RNN thường cần thêm trợ giúp để ổn định việc huấn luyện.
+Với chuỗi độ dài $T$, trong một vòng lặp lan truyền ngược ta tính toán gradient qua $T$ bước thời gian, dẫn đến một chuỗi các tích ma trận có độ phức tạp $\mathcal{O}(T)$.
```
Comment (rootonchair) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:286]:
```suggestion
Với chuỗi độ dài $T$, trong một vòng lặp lan truyền ngược ta tính toán gradient qua $T$ bước thời gian, dẫn đến một chuỗi các tích của ma trận có độ phức tạp $\mathcal{O}(T)$.
```

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 11]
```diff
@@ -276,26 +275,25 @@ predict_ch8('time traveller ', 10, model, vocab, ctx)
...
+Như đã đề cập trong :numref:`sec_numerical_stability`, khi $T$ lớn việc này có thể dẫn đến mất ổn định số học, qua hiện tượng bùng nổ hoặc tiêu biến gradient.
```
Comment (rootonchair) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:287]:
```suggestion
Như đã đề cập trong :numref:`sec_numerical_stability`, khi $T$ lớn việc này có thể dẫn đến mất ổn định số học, biểu hiện qua hiện tượng bùng nổ hoặc tiêu biến gradient.
```

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 12]
```diff
@@ -276,26 +275,25 @@ predict_ch8('time traveller ', 10, model, vocab, ctx)
...
+Do đó, các mô hình RNN thường cần hỗ trợ để ổn định việc huấn luyện.
```
Comment (rootonchair) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:288]:
```suggestion
Do đó, các mô hình RNN thường cần một chút hỗ trợ để ổn định việc huấn luyện.
```

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 13]
```diff
@@ -321,11 +319,10 @@ In this case such an approach may appear entirely unwarranted.
 One alternative is to clip the gradients by projecting them back to a ball of a given radius, say $\theta$ via
 -->
 
-Đôi khi gradient có thể khá lớn và thuật toán tối ưu có thể không hội tụ.
-Ta có thể giải quyết vấn đề này bằng cách giảm tốc độ học $\eta$ hoặc bằng một số thủ thuật bậc cao khác.
-Nhưng điều gì sẽ xảy ra nếu chúng ta hiếm khi nhận được gradient lớn?
-Trong trường hợp này, cách tiếp cận như vậy không được bảo đảm hoàn toàn.
-Một cách khác là cắt bớt các gradient bằng cách chiếu chúng trở lại một quả cầu với bán kính $\theta$ thông qua:
+Đôi khi gradient có thể khá lớn và do đó thuật toán tối ưu không hội tụ.
+Vấn đề này có thể được giả quyết bằng cách giảm tốc độ học $\eta$ hoặc sử dụng một số thủ thuật liên quan tới đạo hàm bậc cao hơn.
```
Comment (rootonchair) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:323]:
```suggestion
Vấn đề này có thể được giải quyết bằng cách giảm tốc độ học $\eta$ hoặc sử dụng một số thủ thuật liên quan tới đạo hàm bậc cao hơn.
```

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 14]
```diff
@@ -337,19 +334,17 @@ Gradient clipping provides a quick fix to the gradient exploding.
...
 
 <!--
 Below we define a function to clip the gradients of a model that is either a `RNNModelScratch` instance or a Gluon model.
 Also note that we compute the gradient norm over all parameters.
 -->
 
-Dưới đây, chúng tôi định nghĩa một hàm để cắt bớt các gradient của mô hình là một thực thể `RNNModelScratch` hoặc một mô hình Gluon.
-Cũng lưu ý rằng chúng tôi tính toán trung bình gradient trên tất cả các tham số.
+Dưới đây, ta định nghĩa hàm gọt gradient, dùng cho cả mô hình lập trình từ đầu `RNNModelScratch` và mô hình tạo bằng Gluon.
+Lưu ý rằng ta tính toán chuẩn của gradient trên tất cả các tham số.
```
Comment (rootonchair) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:347]:
```suggestion
Lưu ý rằng ta tính chuẩn của gradient trên tất cả các tham số.
```

**Pull request merged**